File: examples/put_get.js
var http = require('http');

// call endpoint with params and run callback when finished
function rpc(endpoint, params, callback) {
  var data = '';
  var content = JSON.stringify(params);
  var req = http.request({
    hostname: 'localhost',
    port: 9192,
    headers: {
      'Content-Length': content.length,
      'Accept': 'application/json',
    },
    path: '/rpc/DHash.' + endpoint,
    method: 'POST',
  }, function(res) {
    res.setEncoding('utf8');
    res.on('data', function(d) {
      data += d;
    });
    res.on('end', function() {
      callback(JSON.parse(data));
    });
    res.on('close', function() {
      callback(JSON.parse(data));
    });
  });
  req.on('error', function(e) {
    console.log('problem with request: ' + e.message);
  });
  req.write(content);
  req.end();
};

// a fake user
var user = {
  email: 'mail@domain.tld',
  password: 'so secret',
  name: 'john doe',
};

// insert it into the database
rpc('Put', { 
  Key: new Buffer(user.email).toString('base64'),
  Value: new Buffer(JSON.stringify(user)).toString('base64'),
}, function() {
  // and fetch it again
  rpc('Get', { Key: new Buffer(user.email).toString('base64') }, function(data) {
    console.log('stored and found', JSON.parse(new Buffer(data.Value, 'base64').toString('utf-8')));
  });
});
// output: stored and found { email: 'mail@domain.tld',
// output: password: 'so secret',
// output: name: 'john doe' }

File: examples/put_get.go
package main

import (
  "fmt"
  "github.com/zond/god/client"
  "github.com/zond/god/common"
  "github.com/zond/god/murmur"
)

// a make believe user data structure
type User struct {
  Email    string
  Password string
  Name     string
}

func main() {
  // connect to the default local server
  conn := client.MustConn("localhost:9191")
  // create a user
  user := User{
    Email:    "mail@domain.tld",
    Password: "so secret",
    Name:     "john doe",
  }
  // serialize the user
  bytes := common.MustJSONEncode(user)
  // and put it in the database
  conn.Put(murmur.HashString(user.Email), bytes)
  // try to fetch the user again
  data, _ := conn.Get(murmur.HashString(user.Email))
  var found User
  // to unserialize it
  common.MustJSONDecode(data, &found)
  fmt.Printf("stored and found %+v\n", found)
}

// output: stored and found {Email:mail@domain.tld Password:so secret Name:john doe}
